packages/metadata/src/decorator-factory.ts

```diff
@@ -0,0 +1,539 @@
+> {
+  /**
+   * A symbol to reference the target of a decoration
+   */
+  static TARGET = Symbol('decorationTarget');
+
+  /**
+   * Construct a new class decorator factory
+   * @param key Metadata key
+   * @param spec Metadata object from the decorator function
+   * @param options Options for the decorator. Default to
+   * `{allowInheritance: true}` if not provided
+   */
+  constructor(
+    protected key: string,
+    protected spec: T,
+    protected options?: DecoratorOptions,
+  ) {
+    this.options = Object.assign({allowInheritance: true}, options);
+  }
+
+  protected allowInheritance(): boolean {
+    return this.options && this.options.allowInheritance;
+  }
+
+  /**
+   * Inherit metadata from base classes. By default, this method merges base
+   * metadata into the spec if `allowInheritance` is set to `true`. To customize
+   * the behavior, this method can be overridden by sub classes.
+   *
+   * @param baseMeta Metadata from base classes for the member
+   */
+  protected inherit(baseMeta: T | undefined | null): T {
+    if (!this.allowInheritance()) return this.spec;
+    if (baseMeta == null) return this.spec;
+    if (this.spec == undefined) return baseMeta;
+    if (typeof baseMeta !== 'object') return this.spec;
+    if (Array.isArray(baseMeta) || Array.isArray(this.spec)) {
+      // For arrays, we don't merge
+      return this.spec;
+    }
+    return Object.assign(baseMeta, this.spec);
+  }
+
+  /**
+   * Get name of a decoration target
+   * @param target Class or prototype of a class
+   * @param member Optional property/method name
+   * @param descriptorOrIndex Optional method descriptor or parameter index
+   */
+  getTargetName(
+    target: Object,
+    member?: string | symbol,
+    descriptorOrIndex?: TypedPropertyDescriptor<any> | number,
+  ) {
+    let name =
+      target instanceof Function
+        ? target.name
+        : target.constructor.name + '.prototype';
+    if (member == null && descriptorOrIndex == null) {
+      return 'class ' + name;
+    }
+    if (member == null) member = 'constructor';
+    if (typeof descriptorOrIndex === 'number') {
+      // Parameter
+      name =
+        'parameter ' +
+        name +
+        '.' +
+        member.toString() +
+        '[' +
+        descriptorOrIndex +
+        ']';
+    } else if (descriptorOrIndex != null) {
+      name = 'method ' + name + '.' + member.toString();
+    } else {
+      name = 'property ' + name + '.' + member.toString();
+    }
+    return name;
+  }
+
+  /**
+   * Get the number of parameters for a given constructor or method
+   * @param target Class or the prototype
+   * @param member Method name
+   */
+  getNumberOfParameters(target: Object, member?: string | symbol) {
+    if (target instanceof Function && member == null) {
+      // constructor
+      return target.length;
+    } else {
+      // target[member] is a function
+      return (<any>target)[member!].length;
```

Comment:
Perhaps `{[methodName: string]: Function}` would be a more appropriate type than `any`? I think the type of "a n object with arbitrary methods" may be useful in other places too and therefore is worth sharing via an explicit `type`. (We are using it for Controllers for example.)